src/Data/List/Base.agda

```diff
@@ -188,6 +188,22 @@ unfold B f {n = suc n} s with f s
 ------------------------------------------------------------------------
 -- Operations for deconstructing lists
 
+-- Note that although these combinators can be useful for programming, when
+-- proving it is often a better idea to manually destruct a list argument:
+-- each branch of the pattern-matching will have a refined type.
+
+module _ {a} {A : Set a} where
+
+  uncons : List A → Maybe (A × List A)
+  uncons []       = nothing
+  uncons (x ∷ xs) = just (x , xs)
+
+  head : List A → Maybe A
+  head = Maybe.map proj₁ ∘ uncons
```

Comment:
Is there an advantage in defining it in this composite way over the simple pattern matching? Is it more efficient? Because I feel the standard definition is much easier to understand, and might be more efficient as well? It would also require significantly less imports...

Reply:
Fair enough.